charts/tfy-karpenter-config/files/al2-soci-gpu-provisioner-userdata.sh

```diff
@@ -0,0 +1,169 @@
+    echo "Writing NVreg_EnableGpuFirmware=0 to /etc/modprobe.d/nvidia-gsp.conf"
+    echo "options nvidia NVreg_EnableGpuFirmware=0" | tee --append /etc/modprobe.d/nvidia-gsp.conf
+    echo "Running dracut"
+    dracut -f
+}
+
+# This needs the driver to be already loaded
+# Which is not the case with new EKS AMIs that use Open Kernel Module for NVIDIA drivers
+if [ -f /proc/driver/nvidia/version ]; then
+    if grep "Open Kernel Module" /proc/driver/nvidia/version; then
+        echo "NVIDIA driver is using the open source kernel module. Skipping disabling GSP."
+    else
+        echo "NVIDIA driver is using proprietary NVIDIA kernel module. Disabling GSP."
+        disable_nvidia_gsp
+    fi
+else
+    echo "/proc/driver/nvidia/version does not exist. Skipping disabling GSP."
+fi
+
+# SOCI
+
+CONTAINERD_VERSION_OUTPUT="$(containerd --version)"
+IFS=" " read -a  CONTAINERD_VERSION_STRING <<< "${CONTAINERD_VERSION_OUTPUT}"
+CONTAINERD_VERSION=${CONTAINERD_VERSION_STRING[2]}
+CONTAINERD_CONFIG_FILEPATH="/etc/eks/containerd/containerd-config.toml"
+BACKUP_CONTAINERD_CONFIG_FILEPATH="$CONTAINERD_CONFIG_FILEPATH.bak"
+KUBELET_CONFIG_FILEPATH="/etc/kubernetes/kubelet/kubelet-config.json"
+BACKUP_KUBELET_CONFIG_FILEPATH="/etc/kubernetes/kubelet/kubelet-config.json.bak"
+SOCI_RELEASE_VERSION="0.7.0"
+SOCI_TAR_CHECKSUM="8766cdd479272dcc86299e70a0f7a9343f940c98285c1491bb3c3cdc05b26f47"
+SOCI_TAR_FILENAME="soci-snapshotter-$SOCI_RELEASE_VERSION-linux-amd64.tar.gz"
+SOCI_CONFIG_DIR="/etc/soci-snapshotter-grpc/"
+SOCI_CONFIG_FILEPATH="$SOCI_CONFIG_DIR/config.toml"
+SOCI_SYSTEMD_SERVICE_FILEPATH="/etc/systemd/system/soci-snapshotter.service"
+cp $CONTAINERD_CONFIG_FILEPATH $BACKUP_CONTAINERD_CONFIG_FILEPATH
+cp $KUBELET_CONFIG_FILEPATH $BACKUP_KUBELET_CONFIG_FILEPATH
+
+setup_soci() {( set -ex
+
+yum install fuse -y
+modprobe fuse
+
+curl \
+--silent \
+--show-error \
+--retry 3 \
+--retry-delay 1 \
+-L https://github.com/awslabs/soci-snapshotter/releases/download/v${SOCI_RELEASE_VERSION}/soci-snapshotter-${SOCI_RELEASE_VERSION}-linux-amd64.tar.gz \
+-o $SOCI_TAR_FILENAME
+echo "$SOCI_TAR_CHECKSUM $SOCI_TAR_FILENAME" | sha256sum -c
+
+tar -C /usr/local/bin -xvf $SOCI_TAR_FILENAME soci soci-snapshotter-grpc
+rm $SOCI_TAR_FILENAME
+
+cat > $SOCI_SYSTEMD_SERVICE_FILEPATH << EOF
+[Unit]
+Description=soci snapshotter containerd plugin
+Documentation=https://github.com/awslabs/soci-snapshotter
+After=network.target
+Before=containerd.service
+
+[Service]
+Type=notify
+ExecStart=/usr/local/bin/soci-snapshotter-grpc
+Restart=always
+RestartSec=5
+
+[Install]
+WantedBy=multi-user.target
+EOF
+
+mkdir -p $SOCI_CONFIG_DIR
+
+cat > $SOCI_CONFIG_FILEPATH << EOF
+[cri_keychain]
+enable_keychain=true
+image_service_path="/run/containerd/containerd.sock"
+EOF
+
+systemctl daemon-reload
+systemctl enable --now soci-snapshotter
+systemctl status soci-snapshotter
+
+echo patching $CONTAINERD_CONFIG_FILEPATH
+cat > $CONTAINERD_CONFIG_FILEPATH << EOF
+version = 2
```

Comment:
Have we extracted this from the latest al2 instance? I believe that config also comes with a default nvidia runtime, which I see is missing here.

It would not matter ultimately because gpu-operator's container toolkit with patch and add/update nvidia runtime
But I want to make sure we are not missing any cgroup settings

Reply:
> Have we extracted this from the latest al2 instance?

Yes.

Reply:
```
[root@ip-10-2-51-206 /]# cat /etc/eks/containerd/containerd-config.toml.bak
version = 2
root = "/var/lib/containerd"
state = "/run/containerd"
imports = ["/etc/containerd/config.d/*.toml"]

[grpc]
address = "/run/containerd/containerd.sock"

[plugins."io.containerd.grpc.v1.cri".containerd]
default_runtime_name = "runc"
discard_unpacked_layers = true

[plugins."io.containerd.grpc.v1.cri"]
sandbox_image = "SANDBOX_IMAGE"

[plugins."io.containerd.grpc.v1.cri".registry]
config_path = "/etc/containerd/certs.d:/etc/docker/certs.d"

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
runtime_type = "io.containerd.runc.v2"

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
SystemdCgroup = true

[plugins."io.containerd.grpc.v1.cri".cni]
bin_dir = "/opt/cni/bin"
conf_dir = "/etc/cni/net.d"
```